src/Microsoft.AspNetCore.Sockets.Client/Connection.cs

```diff
@@ -47,20 +47,32 @@ public Connection(Uri url, ILoggerFactory loggerFactory)
         public Task StartAsync(HttpClient httpClient) => StartAsync(transport: null, httpClient: httpClient);
         public Task StartAsync(ITransport transport) => StartAsync(transport: transport, httpClient: null);
 
-        // TODO HIGH: Fix a race when the connection is being stopped/disposed when start has not finished running
-        public async Task StartAsync(ITransport transport, HttpClient httpClient)
+        public Task StartAsync(ITransport transport, HttpClient httpClient)
         {
-            _transport = transport ?? new WebSocketsTransport(_loggerFactory);
+            _startTask = StartAsyncInternal(transport, httpClient);
+            return _startTask;
+        }
 
+        private async Task StartAsyncInternal(ITransport transport, HttpClient httpClient)
+        {
             if (Interlocked.CompareExchange(ref _connectionState, ConnectionState.Connecting, ConnectionState.Initial)
                 != ConnectionState.Initial)
             {
                 throw new InvalidOperationException("Cannot start a connection that is not in the Initial state.");
             }
 
+            _transport = transport ?? new WebSocketsTransport(_loggerFactory);
```

Comment:
No, you are right - it should be done as close to starting the transport as possible. I will fix this.